%\VignetteIndexEntry{erccdashboard examples}

\documentclass{article}
\usepackage{fullpage}
\begin{document}
\SweaveOpts{concordance=TRUE}
\SweaveOpts{keep.source=TRUE}
\SweaveOpts{width=7, height=7}
\DefineVerbatimEnvironment{Sinput}{Verbatim} {xleftmargin=1em}
\DefineVerbatimEnvironment{Soutput}{Verbatim}{xleftmargin=1em}
\DefineVerbatimEnvironment{Scode}{Verbatim}{xleftmargin=1em}
\fvset{listparameters={\setlength{\topsep}{0pt}}}
\renewenvironment{Schunk}{\vspace{\topsep}}{\vspace{\topsep}}
<<echo=false>>=
options(width=60, continue = "  ")
#options(SweaveHooks=list(fig=function()
#                par(mar=c(5.1,4.1,1.1,2.1))))
library( "erccdashboard" )
@    
\title{{\tt erccdashboard} Package Vignette}
\author{Sarah A. Munro}
\maketitle

This vignette describes the use of the erccdashboard R package to analyze 
External RNA Control Consortium (ERCC) spike-in control ratio mixtures in gene
expression experiments. If you use this package for method validation of your
gene expression experiments please cite our manuscript that describes this R 
package.

In this vignette we demonstrate analysis of two types of gene expression experiments
from the SEQC project that used ERCC control ratio mixture spike-ins:
\begin{itemize}
  \item Rat toxicogenomics methimazole-treated and control samples
  \item Human reference RNA samples from the MAQC I project, Universal Human 
  Reference RNA (UHRR) and Human Brain Reference RNA (HBRR)
\end{itemize}

A subset of the large data set produced in the SEQC study are provided here as 
examples. The three sets of example data are: 

\begin{enumerate}
  \item Rat toxicogenomics RNA-Seq gene expression count data
  \item UHRR/HBRR RNA-Seq gene expression count data
  \item UHRR/HBRR Microarray gene expression fluorescent intensity data
\end{enumerate}


\section{Rat Toxicogenomics Example: MET (methimazole treatment) and CTL 
(control) Experiment}
\subsection{Load data and define input parameters} 

Load the package gene expression data.
<<loadExampleData, keep.source=TRUE>>=
data(SEQC.Example)
@

The R workspace should now contain 5 objects
Three of these objects are gene expression experiment expression measures:
\begin{itemize}
  \item Lab10.array.UHRR.HBRR - Fluorescent signal data from an Illumina 
        beadarray microarray experiment with UHRR and HBRR
  \item COH.RatTox.ILM.MET.CTL.countTable - RNA-Seq count data from a
        rat toxicogenomics experiment
  \item Lab5.ILM.UHRR.HBRR.countTable - RNA-Seq count data from Lab 5 in the SEQC
        interlaboratory study with UHRR and HBRR
\end{itemize}
The other two objects are vectors of total reads for the 2 sequencing experiments
\begin{itemize}
  \item COH.RatTox.ILM.MET.CTL.totalReads - total sequenced reads factors for 
        each column in the corresponding rat experiment count table
  \item Lab5.ILM.UHRR.HBRR.totalReads - total sequenced reads factors for 
        each column in the corresponding UHRR/HBRR count table
\end{itemize}


\subsection{Quick analysis: runDashboard}
To run the default analysis function runDashboard on the MET-CTL rat 
toxicogenomics RNA-Seq experiment, the following input arguments are required:

<<defineInputData,keep.source=TRUE>>=
datType = "count" # "count" for RNA-Seq data, "array" for microarray data
isNorm = FALSE # flag to indicate if input expression measures are already
               # normalized, default is FALSE 
exTable = COH.RatTox.ILM.MET.CTL.countTable # the expression measure table
filenameRoot = "COH.ILM" # user defined filename prefix for results files
sample1Name = "MET" # name for sample 1 in the experiment
sample2Name = "CTL" # name for sample 2 in the experiment
erccmix = "RatioPair" # name of ERCC mixture design, "RatioPair" is default
erccdilution = 1/100 # dilution factor used for Ambion spike-in mixtures
spikeVol = 1 # volume (in microliters) of diluted spike-in mixture added to 
             #   total RNA mass
totalRNAmass = 0.500 # mass (in micrograms) of total RNA 
choseFDR = 0.1 # user defined false discovery rate (FDR), default is 0.05
@

The first input argument, datType, indicates whether that data is integer count 
data from an RNA-Seq experiment ("count") or data from a microarray experiment 
("array"). The isNorm argument indicates if the input expression measures are 
already normalized, the default value is FALSE. If you want to use normalized
RNA-Seq or microarray data in the analysis, the isNorm
argument must be set to TRUE. For array data this will modify the
weights applied in limma DE testing. For RNA-Seq data, DE testing results must 
be available in the working directory in a file named "filenameRoot ERCC Pvals.csv"

The third argument, exTable, is the expression measure table.

Take a look at the RatTox experiment count table.
<<inspectRatCount,keep.source=TRUE>>=
head(COH.RatTox.ILM.MET.CTL.countTable)
@
The first column of the expression measure table, Feature, contains unique names for all
the transcripts that were quantified in this experiment. The remaining columns 
represent replicates of the pair of samples, in this expression measure table 
the control sample is labeled CTL and the treatment sample is labeled MET. An 
underscore is included to separate the sample names from the replicate numbers 
during analysis. This column name format Sample\textunderscore{}Rep is required 
for the columns of any input expression measure table. Only one underscore 
(\textunderscore{}) should be used in the column names.

The default differential expression testing of
RNA-Seq experiments in the erccdashboard is done with the QuasiSeq package, 
which requires the use of integer count data. The default normalization of the data
is 75th percentile (also known as upper quartile) normalization. 
It is optional to provide a vector of per replicate normalization factors 
through the input argument repNormFactor, such as a vector of total reads for 
each replicate. The example total reads vectors we provide here were derived from
the FASTQ files associated with each column in the RNA-Seq experiment count tables.
Any repNormFactor vector will be used as a library size normalization factor for each
column of exTable. This will be adjusted to be a per million reads factor.

For any experiment the sample spiked with ERCC Mix 1 is sample1Name and the 
sample spiked with ERCC Mix 2 is sample2Name. In this experiment 
sample1Name = MET and sample2Name = CTL. For a more robust experimental design 
the reverse spike-in design could be created using additional replicates of the
treatment and control samples. ERCC Mix 2 would be spiked into MET samples and 
ERCC Mix 1 would be spiked into CTL control replicates. 

The dilution factor of the pure Ambion ERCC mixes prior to spiking into total 
RNA samples is erccdilution. The amount of diluted ERCC mix spiked into
the total RNA sample is spikeVol (units are $\mu$L). The mass of total RNA 
spiked with the diluted ERCC mix is totalRNAmass (units are $\mu$g).

The final required input parameter, choseFDR, is the False Discovery Rate (FDR) 
for differential expression testing. A typical choice would be 0.05 (5\% FDR), 
so this is the default choseFDR value. For the rat data sets a more liberal FDR 
was used, choseFDR = 0.1.

The function runDashboard.R is provided for convenient default erccdashboard 
analysis. Execution of the runDashboard function calls the default functions for
erccdashboard analysis and reports parameters and progress to the R console. The
functions called within runDashboard.R are also available to the user (details 
provided in Section 4). 

All data and analysis results are stored in the list object exDat. For convenience 
the main diagnostic figures are saved to a pdf file and the exDat object is 
saved to an .RData object named using the filenameRoot provided by the user.

Use the following command to run the default runDashboard script:

<<runDashboardSEQCcount, keep.source=TRUE>>=
exDat <- runDashboard(datType = "count", isNorm = FALSE,
                       exTable = COH.RatTox.ILM.MET.CTL.countTable,
                       filenameRoot = "COH.ILM",sample1Name = "MET",
                       sample2Name = "CTL",erccmix = "RatioPair",
                       erccdilution = 1/100, spikeVol = 1,
                       totalRNAmass = 0.500, choseFDR = 0.1)
@

\subsection{Results of dashboard analysis}
The summary function will give a top level view
of the exDat list structure. The str function will give more detail. It is a good idea
to set the max.level argument in the str function, because by the end of the 
analysis the exDat structure is quite large.
<<initializeData,keep.source=TRUE>>=
summary(exDat)
@
The figures from the analysis are stored in \verb|exDat$Figures|. The four main
diagnostic figures that are saved to pdf are the dynRangePlot, rocPlot, 
lodrERCCPlot, and maPlot.
\clearpage
\begin{center}
<<ratPlotA,fig=TRUE >>=
exDat$Figures$dynRangePlot
@
\end{center}
For this particular experiment the relationship between abundance and signal for
the ERCC controls show that the measurement results span a 2$^{15}$ dynamic range.
These ERCC mixtures were designed to span a 2$^{20}$ dynamic range, but there was 
insufficient evidence to reliably quantify ERCC transcripts at low 
abundances.
\clearpage
\begin{center}
<<ratPlotB,fig=TRUE >>=
exDat$Figures$rocPlot
@
\end{center}
The receiver operator characteristic (ROC) curve and the Area Under the Curve (AUC)
statistic provide evidence of the diagnostic power for detecting differential expression
in this rat toxicogenomics experiment. As expected with increased fold change,
diagnostic power increases. The AUC summary statistic for different experiments
can be used to compare diagnostic performance.
\clearpage
\begin{center}
<<ratPlotC,fig=TRUE >>=
exDat$Figures$lodrERCCPlot
@
\end{center}
By modeling the relationship between average signal and p-values we can obtain 
Limit of Detection of Ratios (LODR) estimates for each differential fold change 
(or Ratio,indicated by color) and a threshold p-value, p.thresh, indicated by 
the dotted black line. LODR values can be compared between
experiments to evaluate the ability to detect differences between samples as a function
of transcript abundance.
\clearpage
\begin{center}
<<ratPlotD,fig=TRUE >>=
exDat$Figures$maPlot
@
\end{center}
An MA plot (Ratio of Signals vs Average Signals) shows the ratio measurements of
transcripts in the pair of samples as a function of abundance. The ERCC control 
ratios measurements are coded to indicate which controls are above a given LODR 
(solid circles) or below the LODR (open circles). This plot also shows the v
ariability in ratio measurements as a function of dynamic range and the bias in 
control ratio measurements (r$_m$), which is influenced by the mRNA fraction difference
between the pair of samples.

\section{SEQC Reference RNA (UHRR and HBRR) Examples}
All UHRR and HBRR experiments for the SEQC interlaboratory and interplatform 
analysis were performed with aliquots from single large samples of UHRR and 
HBRR. The count data and microarray data provided here are from experiments 
using the same original UHRR and HBRR samples.

\subsection{RNA-Seq UHRR vs. HBRR experiment analysis}
<<SEQCMainCount keep.source=TRUE>>=
exDat <- runDashboard(datType="count", isNorm = FALSE,
                      exTable=Lab5.ILM.UHRR.HBRR.countTable,
                      filenameRoot="Lab5", sample1Name = "UHRR", 
                      sample2Name = "HBRR", erccmix = "RatioPair",
                      erccdilution = 1, spikeVol = 50, 
                      totalRNAmass = 2.5*10^(3), choseFDR = 0.01)
@
\clearpage
\begin{center}
<<seqcCountPlotA,fig=TRUE >>=
exDat$Figures$dynRangePlot
@
\end{center}
Compared with the rat toxicogenomics experiment the UHRR/HBRR experiment from Lab 5 
captures the full dynamic range of 2$^{20}$ in the ERCC mixture design. This 
difference can be attributed to an much greater sequencing depth in the 
UHRR/HBRR experiments at each laboratory (full fluidic capacity of an 
instrument) compared to the MET-CTL rat toxicogenomics experiment. This 
difference is apparent in the total reads vectors for the two different 
experiments:
<<totalReadsCompare>>=
COH.RatTox.ILM.MET.CTL.totalReads
Lab5.ILM.UHRR.HBRR.totalReads
@
\clearpage
\begin{center}
<<seqcCountPlotB,fig=TRUE >>=
exDat$Figures$rocPlot
@
\end{center}

Given the sequencing depth of the Lab 5 UHRR/HBRR experiment it is unsurprising that 
the ROC curve results show that almost all of the spiked ERCC transcripts were 
detected and tested for differential expression. Again we see that the diagnostic
power decreases as the differential ratio decreases.At first glance, one may attempt
to compare these ROC curve and AUC results to the rat toxicogenomics results, but
this is ill-advised because the experiments are different. Not only
is the sequencing depth different for the two experiments, but the UHRR/HBRR 
sample transcripts also have high levels of differential expression, whereas 
the rat MET/CTL samples have much lower levels of differential expression. This
difference in differential expression is shown in the spread and density of the 
points representing the endogenous transcripts in the MA plots for each experiment.
 
\clearpage
\begin{center}
<<seqcCountPlotC,fig=TRUE >>=
exDat$Figures$lodrERCCPlot
@
\end{center}

Much lower LODR estimates are found in the Lab 5 UHRR/HBRR experiment compared to the 
rat experiment. One can also see that the p-values in this analysis are much smaller
than those in the rat experiment due to the large amount of differentially expressed
transcripts in the UHRR/HBRR comparison.
\clearpage
\begin{center}
<<seqcCountPlotD,fig=TRUE >>=
exDat$Figures$maPlot
@
\end{center}

Variability in the ratio measurements in this experiment is much lower than in
the rat experiment. This is due to the spike-in design of the two experiments.
In the rat experiment the 3 replicates for each condition represent biological 
replicates and every replicate was spiked separately with Ambion mix. In the 
UHRR/HBRR experiment a large batch of each type of reference RNA was spiked once
and sample aliquots were then distributed for measurement.
The bias in control ratios in this experiment is much greater than what was observed in
the rat experiment. This bias is likely due to the known mRNA fraction difference
between the UHRR and HBRR RNA samples.

\clearpage
\subsection{Microarray UHRR vs. HBRR experiment analysis}
Unnormalized fluorescent signals are expected for erccdashboard analysis of 
microarray data, the data will be log2 transformed for DE testing with the limma
package. The repNormFactor variable should
be NULL for microarray analysis, and repNormFactor is set to NULL if it is 
missing in the list of arguments for the runDashboard function. Each array 
replicate is normalized by the 75th quantile fluoresence intensity 
value of that array.
<<SEQCMainArray, keep.source=TRUE>>=
exDat <- runDashboard(datType="array", isNorm = F,
                      exTable=Lab10.array.UHRR.HBRR,
                      filenameRoot = "Lab10.array",
                      sample1Name = "UHRR", sample2Name="HBRR",
                      erccmix = "RatioPair", erccdilution = 1, 
                      spikeVol = 50, totalRNAmass = 2.5*10^(3), choseFDR=0.01)
@
\clearpage
\begin{center}
<<seqcArrayPlotA,fig=TRUE >>=
exDat$Figures$dynRangePlot
@
\end{center}

The dynamic range of the microarray experiment on reference RNA samples was 2$^{10}$.
This figure shows the signal compression for high abundance transcripts.
\clearpage
\begin{center}
<<seqcArrayPlotB,fig=TRUE >>=
exDat$Figures$rocPlot
@
\end{center}

Comparison of these microarray ROC curves and AUC statistics to the results in the RNA-Seq 
UHRR/HBRR experiment from Lab 5 show similar diagnostic power 
for the two platforms. Fewer controls were detected above background in the microarray experiment, and this should be considered when comparing the diagnostic performance
of the two platforms. More stringent filtering of RNA-Seq data would likely lead to improved diagnostic
performance.
\clearpage
\begin{center}
<<seqcArrayPlotC,fig=TRUE >>=
exDat$Figures$lodrERCCPlot
@
\end{center}

LODR estimates for microarray data are derived from the intersection with the 
threshold p-value at the lowest signal. The same FDR was chosed for the RNA-Seq 
and microarray measurements of the UHRR/HBRR samples, it is interesting to note 
that different threshold p-values were derived for each type of experiment using
the same FDR = 0.01, p.thresh = 0.1518 for RNA-Seq at Lab 5 and 
p.thresh = 0.0430 for the microarray experiment.
\clearpage
\begin{center}
<<seqcArrayPlotD,fig=TRUE >>=
exDat$Figures$maPlot
@
\end{center}

The variability and bias in the control ratio measurements 
in the microarray data measurement of UHRR/HBRR are similar to the observations in
the RNA-Seq UHRR/HBRR experiment.

\section{Comparison of Performance Between Experiments}
The performance metrics provided here derived from measurements of ERCC ratios in
gene expression experiments (AUC, LODR, r$_m$, and the standard deviations of 
the ERCC ratio measurements) can be used to assess performance between experiments within
the same laboratory, or between different laboratories or technology platforms.

\section{Analysis Details: Advanced Use of Functions in runDashboard}
The analysis functions contained in the convenience wrapper function 
\verb|runDashboard| can also be used directly by the user.
Comments are provided above each analysis step included in \verb|runDashboard| to
describe the purpose and ordering constraints.
View the runDashboard script to see comments describing the analysis functions and the ordering constraints:
<<viewDashboardOrder>>=
runDashboard
@

\subsection{Flexibility in Differential Expression Testing}
The \verb|geneExprTest| function wraps the QuasiSeq differential expression 
testing package for datType = "count" or uses the limma package for differential
expression testing if datType = "array". The function uses the DE testing 
p-value results and choseFDR parameter to select a threshold p-value for LODR 
estimation.
In the case of count data if QuasiSeq DE testing has already been done and 
a correctly formatted quasiseq results csv file is provided with the
necessary DE test results, then geneExprTest will have 
reduced runtime. The function will look for a csv file with the name 
"filenameRoot.quasiSeq.res.csv" and columns with names "Feature", "pvals", and 
"qvals" must be in the file.

To use results from another DE testing tool (instead of QuasiSeq) a csv file with
the name "filenameRoot ERCC pvals.csv" and correct column headers should be in the working directory.
For RNA-Seq ("count") data, this ERCC pvals file is required to complete analysis without 
QuasiSeq DE testing or if the input data (exTable) is already normalized 
(isNorm = TRUE).
If isNorm is TRUE, then the software asks for user input about length normalization.
Type Y at the command line in the R console if the data is length normalized (e.g. FPKM or RPKM data) otherwise type N.

\subsection{Options for LODR Estimation}
The default behavior of runDashboard is to use the \verb|estLODR| function to
obtain an LODR estimate using empirical data from the ERCCs and a model-based 
simulation using the endogenous genes in the sample. The type of LODR estimation
is selected using the argument kind in the \verb|estLODR| function. The other 
parameter that may be adjusted is the probability for the LODR estimate, in the 
default analysis prob = 0.9 is selected.

\subsection{Options for Printing Plots to File}
The function savePlots will save selected figures to a pdf file. The default 
is the 4 manuscript figures to a single page (plotsPerPg = "manuscript"). 
If plotsPerPg = "single" then each plot is placed on an 
individual page in one pdf file. If plotlist is not defined (plotlist = NULL)
then all plots in \verb|exDat$Figures| are printed to the file.

To print 4 plots from manuscript to a single page pdf file use 
plotsPerPg = "manuscript" in the \verb|saveERCCPlots| function or to create a 
multiplepage pdf of all plots produced use plotsPerPg = "single". 

\subsection{Analysis of Alternative Spike-in Designs}
By default the package is configured to analyze the ERCC ratio mixtures produced
by Ambion (ERCC ExFold RNA Spike-In Mixes, Catalog Number 4456739). This pair of
control ratio mixtures were designed to have 1:1, 4:1, 1:1.5, and 1:2 ratios of 92
distinct RNA transcripts (23 different RNA control sequences are in each of these 
four ratio subpools). Alternative ERCC RNA control ratio mixture designs can be 
produced using the NIST DNA Plasmid Library for External Spike-in Controls (NIST 
Standard Reference Material 2374, 
https://www-s.nist.gov/srmors/certificates/2374.pdf). For example, a pair of RNA
control mixtures could be created with a ternary ratio design, three subpools of 
RNA controls with either no change (1:1) or 2-fold increased (2:1) and 2-fold
decreased (1:2) relative abundances between the pair of mixtures (Mix 1/Mix 2).
To use alternative spike-in mixture designs with the dashboard a csv file must
be provided to the package with the argument userMixFile for the initDat function. 

If all samples from both conditions were only spiked with a single ERCC mixture 
(e.g. Ambion Catalog Number 4456740, ERCC RNA Spike-In Mix) a limited subset of 
the package functions can be used (\verb|initDat|, \verb|est_r_m|, and 
\verb|dynRangePlot|. For \verb|initDat| use ERCCMixes="Single" and \verb|est_r_m| and 
\verb|dynRangePlot| functions can then be used to examine the mRNA fraction 
differences for the pair of samples and evaluate the dynamic range of the 
experiment. 

\section{Notes on R version and session information}
The results shown in this R vignette are the same as the results shown in our
manuscript and were obtained in R version 3.0.2 with the following session 
information.
<<sessionData>>=
sessionInfo()
@
 
\end{document}